component/pyroscope/ebpf/ebpfspy/session.go

```diff
@@ -0,0 +1,275 @@
+	s.modMutex.Lock()
+	defer s.modMutex.Unlock()
+
+	s.roundNumber += 1
+
+	keys, values, batch, err := s.getCountsMapValues()
+	if err != nil {
+		return fmt.Errorf("get counts map: %w", err)
+	}
+
+	type sf struct {
+		pid    uint32
+		count  uint32
+		kStack []byte
+		uStack []byte
+		comm   string
+		labels *sd.Target
+	}
+	var sfs []sf
+	knownStacks := map[uint32]bool{}
+	for i := range keys {
+		ck := &keys[i]
+		value := values[i]
+
+		if ck.UserStack >= 0 {
+			knownStacks[uint32(ck.UserStack)] = true
+		}
+		if ck.KernStack >= 0 {
+			knownStacks[uint32(ck.KernStack)] = true
+		}
+		labels := s.serviceDiscovery.FindTarget(ck.Pid)
+		if labels == nil {
+			continue
+		}
+		uStack := s.getStack(ck.UserStack)
+		kStack := s.getStack(ck.KernStack)
+		sfs = append(sfs, sf{
+			pid:    ck.Pid,
+			uStack: uStack,
+			kStack: kStack,
+			count:  value,
+			comm:   getComm(ck),
+			labels: labels,
+		})
+	}
+
+	sb := stackBuilder{}
+	for _, it := range sfs {
+		sb.rest()
+		sb.append(it.comm)
+		s.walkStack(&sb, it.uStack, it.pid)
+		s.walkStack(&sb, it.kStack, 0)
+		reverse(sb.stack)
+		err = cb(it.labels, sb.stack, uint64(it.count), it.pid)
+		if err != nil {
+			return err
+		}
+	}
+	if err = s.clearCountsMap(keys, batch); err != nil {
+		return fmt.Errorf("clear counts map %w", err)
+	}
+	if err = s.clearStacksMap(knownStacks); err != nil {
+		return fmt.Errorf("clear stacks map %w", err)
+	}
+	return nil
+}
+
+func (s *Session) Stop() {
+	for _, pe := range s.perfEvents {
+		_ = pe.Close()
+	}
+	s.bpf.Close()
+}
+
+func (s *Session) initArgs() error {
+	var zero uint32
+	var tgidFilter uint32
+	if s.pid <= 0 {
+		tgidFilter = 0
+	} else {
+		tgidFilter = uint32(s.pid)
+	}
+	arg := &profileBssArg{
+		TgidFilter: tgidFilter,
+	}
+	if err := s.bpf.Args.Update(&zero, arg, 0); err != nil {
+		return fmt.Errorf("init args fail: %w", err)
+	}
+	return nil
+}
+
+func (s *Session) attachPerfEvents() error {
+	var cpus []uint
+	var err error
+	if cpus, err = cpuonline.Get(); err != nil {
+		return fmt.Errorf("get cpuonline: %w", err)
+	}
+	for _, cpu := range cpus {
+		pe, err := newPerfEvent(int(cpu), int(s.sampleRate))
+		if err != nil {
+			return fmt.Errorf("new perf event: %w", err)
+		}
+		s.perfEvents = append(s.perfEvents, pe)
+
+		err = pe.attachPerfEvent(s.bpf.profilePrograms.DoPerfEvent)
+		if err != nil {
+			return fmt.Errorf("attach perf event: %w", err)
+		}
+	}
+	return nil
+}
+
+func (s *Session) getStack(stackId int64) []byte {
+	if stackId < 0 {
+		return nil
+	}
+	stackIdU32 := uint32(stackId)
+	res, err := s.bpf.profileMaps.Stacks.LookupBytes(stackIdU32)
+	if err != nil {
+		return nil
+	}
+	return res
+}
+
+func (s *Session) walkStack(sb *stackBuilder, stack []byte, pid uint32) {
+	if len(stack) == 0 {
+		return
+	}
+	var stackFrames []string
+	for i := 0; i < 127; i++ {
+		it := stack[i*8 : i*8+8]
+		ip := binary.LittleEndian.Uint64(it)
+		if ip == 0 {
+			break
+		}
+		sym := s.symCache.resolve(pid, ip, s.roundNumber)
```

Comment:
Does this allow a user to distinguish between user space and kernel symbols? The kernel contains a lot of functions. Some of these could easily overlap with symbols from user applications or libraries. If one then uses Pyroscope's Sandwich-View feature, wouldn't these be merged?

Reply:
Also, is there an option to exclude user and/or kernel symbols from the trace?

Reply:
As of right now there is no way to distinguish between user/kernel.
There's no option to exclude user/kernel traces right now, but i think it's useful and i will add them.